BeeSwift/AppDelegate.swift

```diff
@@ -142,4 +126,135 @@ class AppDelegate: UIResponder, UIApplicationDelegate, UNUserNotificationCenterD
             completionHandler([.alert, .sound, .badge])
         }
     }
+    
+    // processing user's response to a notification
+    func userNotificationCenter(_ center: UNUserNotificationCenter, didReceive response: UNNotificationResponse, withCompletionHandler completionHandler: @escaping () -> Void) {
+        
+        guard response.actionIdentifier == UNNotificationDefaultActionIdentifier else {
+            completionHandler()
+            return
+        }
+        
+        guard let slug = response.notification.request.getSlug() else {
+            completionHandler()
+            return
+        }
+        
+        print("found slug: \(slug)")
+        
+        self.fetchGoalsIfNeededThenViewMatchingGoalWith(slug: slug)
+        
+        completionHandler()
+    }
+    
+    // MARK: opening goals
+    
+    private func fetchGoalsIfNeededThenViewMatchingGoalWith(slug: String) {
+        if CurrentUserManager.sharedManager.goals.isEmpty || CurrentUserManager.sharedManager.goalsFetchedAt.timeIntervalSinceNow < -3600 {
+            // apparently no data or stale data
+            
+            CurrentUserManager.sharedManager.fetchGoals(success: { goals in
+                CurrentUserManager.sharedManager.goals = goals
+                self.openGoalWith(slug)
+            }, error: nil)
+            
+        } else {
+            self.openGoalWith(slug)
+        }
+    }
+    
+    /// given a goal we shall navigate to it
+    private func openGoal(goal: JSONGoal) {
+        guard let navigationController = window?.rootViewController as? UINavigationController else { return }
```

Comment:
Elsewhere in code (in `GalleryViewController`) these sort of VC manipulations are taking place. 
How about refactoring such that these types of tasks appear next to each other (for now that GalleryVC perform them)